### DISCUSSION TITLE
Polyfills and reflect-metadata

### DISCUSSION THREAD

**Discussion by LIMPIX31** [vitest-dev/vitest#2646] [Q&A]
How do I set up `vitest` for `in-source` testing, so that I can use `reflect-metadata` and prototype extensions for example?

**Comment by sheremet-va** [comment]
You need to add your source code files to `includeSource` property in your config:

```ts
// vitest.config.ts
export default {
  test: {
    includeSource: ['src/**/utils.ts']
  }
}
```

And use `import.meta.vitest` to define tests:

```ts
export function add(...args: number[]) {
  return args.reduce((a, b) => a + b, 0)
}

if (import.meta.vitest) {
  const { it, expect } = import.meta.vitest
  it('add', () => {
    expect(add()).toBe(0)
    expect(add(1)).toBe(1)
    expect(add(1, 2, 3)).toBe(6)
  })
}
```

Docs: https://vitest.dev/guide/in-source.html

**Reply by LIMPIX31** [reply]
> You need to add your source code files to `includeSource` property in your config:
> 
> ```ts
> // vitest.config.ts
> export default {
>   test: {
>     includeSource: ['src/**/utils.ts']
>   }
> }
> ```
> 
> And use `import.meta.vitest` to define tests:
> 
> ```ts
> export function add(...args: number[]) {
>   return args.reduce((a, b) => a + b, 0)
> }
> 
> if (import.meta.vitest) {
>   const { it, expect } = import.meta.vitest
>   it('add', () => {
>     expect(add()).toBe(0)
>     expect(add(1)).toBe(1)
>     expect(add(1, 2, 3)).toBe(6)
>   })
> }
> ```
> 
> Docs: https://vitest.dev/guide/in-source.html

It doesn't work.
```ts
// src/vitest.env.ts
import 'reflect-metadata'

Number.prototype.map = ...
```

```ts
// vitest.config.ts
export default defineConfig({
  test: {
    includeSource: ['src/**/*.{ts,tsx}', 'src/vitest.env.ts'],
    environment: 'happy-dom',
    coverage: {
      provider: 'istanbul',
    },
  },
})
```

Error:
> this._progress.map is not a function

**Reply by LIMPIX31** [reply]
?

**Reply by shellscape** [reply]
it's a shame this never got a followup. looking for the same answer. importing reflect-metadata with includeSource is also not working for me.

**Reply by yakirgot** [reply]
I did it by creating `test-setup.ts` file and then including it in `setupFiles`.
